File: core/scripts/Java-Library/src/chh/math/set/Pair.java
package chh.math.set;
/**
 * The Axiom of Pair:
 * Given two sets, there exists a set that contains exactly them both.
 * <p>The order of these elements is <strong>not relevant</strong>; that is,
 * {@code (new Pair(a,b)).equals(new Pair(b,a))==true}.</p>
 * @author  <a href="mailto:chrisharvey2pi@gmail.com">Christopher H. Harvey</a>
 * @version 2013.12.08
 */
public class Pair extends Set {
	/** One element of this Pair. */
	private final Set element1;
	/** Another element of this Pair. */
	private final Set element2;
	/**
     * Constructs a new Pair object.
	 * 
	 * The order of parameters is irrelevant.
	 * @param x one element of this set
	 * @param y another element of this set
	 */
    public Pair(Set x, Set y) {
		this.element1 = x;
		this.element2 = y;
	}
	/**
	 * Constructs a new Pair object containing two equal sets.
	 * (a convenience constructor)
	 * @see Singleton
	 * @param x an element of this set (equal to the other element)
	 */
	public Pair(Set x) {
		this(x,x);
	}
	/**
	 * Constructs a new Pair object containing two empty sets.
	 * (a convenience constructor)
	 * @see Singleton
	 * @see EmptySet
	 */
	public Pair() {
		this(new EmptySet());
	}
	/**
     * {@inheritDoc}
	 * A pair contains exactly its elements.
     */
    @Override
	public boolean contains(Set x) {
		return x.equals(this.element1) || x.equals(this.element2);
	}
	/**
     * {@inheritDoc}
     */
	@Override
	public boolean includes(Set x) {
		return x.isEmpty() || x.isSingletonOf(this.element1) || x.isSingletonOf(this.element2) || x.equals(this);
	}
	/**
     * {@inheritDoc}
	 * A Pair is always nonempty.
	 * @return false
     */
	@Override
	public final boolean isEmpty() {
		return false;
	}
	/**
	 * {@inheritDoc}
	 */
	@Override
	public boolean containsEmpty() {
		return this.element1.isEmpty() || this.element2.isEmpty();
	}
	/**
     * {@inheritDoc}
	 * @return true if this Pair's elements are equal
	 */
	@Override
	public boolean isSingleton() {
		return this.element1.equals(this.element2);
	}
	/**
     * {@inheritDoc}
	 * All Pair objects are pairs.
	 * @return true
	 */
	@Override
	public final boolean isPair() {
		return true;
	}
	/**
     * {@inheritDoc}
	 * The power set of x is a pair if x is a singleton or if x is empty.
	 * If so, this pair contains the empty set and x.
	 * @return true if x is a singleton or is empty, and this set contains x and the empty set
	 */
	@Override
	public boolean isPowerSetOf(Set x) {
		boolean p = this.containsEmpty() && this.contains(x); // a fact about all power sets
		boolean e = x.isSingleton() || x.isEmpty();
		return p && e;
	}
	/**
     * {@inheritDoc}
	 * No Pair objects are inductive.
	 * @return false
	 */
	@Override
	public boolean isInductive() {
		return false;
	}
	/**
     * {@inheritDoc}
	 * @return {@code { element1, element2 }}
	 */
	@Override
	public String toString() {
		String s = "{ " + this.element1;
		if (!this.isSingleton()) {
			s += ",  " + this.element2;
		}
		s += " }";
		return s;
	}
}

File: core/scripts/set/Pair.js
/**
  * Given two Sets, there exists a Set that owns exactly them both.
  * The order of these elements is not relevant, that is,
  * `(new Pair(a,b)).equals(new Pair(b,a)) === true`
  * @extends   Set
  * @param `a` one element in this set
  * @param `b` another element in this set
  */
function Pair(a, b) {
  this.element1 = a || new EmptySet();
  this.element2 = b || this.element1;
  Set.call(this, this.element1, this.element2);
}
Util.extend(Pair, Set);

Pair.prototype.owns = function (x) {
  return this.element1.equals(x) || this.element2.equals(x);
}

Pair.prototype.includes = function (x) {
  return x.isEmpty() // every set includes the empty set
         || x.isSingletonOf(this.element1)
         || x.isSingletonOf(this.element2)
         || x.equals(this); // WARNING: CYCLICAL DEFINITIONS
}

Pair.prototype.equals = function (x) {
  boolean sameObject = (this === x); // convenience statement: equal Objects are equal sets (but not vice versa)
  boolean bothEmpty = this.isEmpty() && x.isEmpty(); // convenience statement: all empty sets are equal
  boolean includeEachOther = this.includes(x) && x.includes(this); // definition of set equality
  return sameObject || bothEmpty || includeEachOther;
}

Pair.prototype.toString = function () {
  var s = '{ ' + this.element1;
  if (!this.isSingleton()) s += ',  ' + this.element2;
  s += ' }';
  return s; // `{ element1,  element2 }`
}


Pair.prototype.isEmpty = function () {
  return false;
}

Pair.prototype.ownsEmpty = function () {
  // overrides Set.prototype.ownsEmpty() {return this.owns(new EmptySet());}
  return this.element1.isEmpty() || this.element2.isEmpty();
}

Pair.prototype.isSingleton = function () {
  return this.element1.equals(this.element2);
}

Pair.prototype.isPair = function () {
  return true;
}
Pair.prototype.isPairOf = function (x, y) {
  return this.owns(x) && this.owns(y);
}

Pair.prototype.isPowerSetOf = function (x) {
  // The power set P(x) of any set x owns the empty set, x itself, and possibly more elements.
  // If x is empty, then P(x) is exacly the set that owns x, which makes it a singleton, which is a pair.
  // If x is a singleton, then P(x) owns the empty set and x, and nothing else, which makes it a non-singleton pair.
  // If x owns more than 1 element, then P(x) will own more than 2 elements, so it will not be a pair.
  var p = this.ownsEmpty() && this.owns(x); // a fact about all power sets
  var e = x.isSingleton() || x.isEmpty();
  return p && e;
  // PowerSet.prototype.isPowerSetOf.call(this, x)
}

Pair.prototype.isInductive = function () {
  return false;
}
